src/kommatablemodel.cpp

```diff
@@ -2,10 +2,11 @@
 
 KommaTableModel::KommaTableModel(QList <QStringList> string_matrix, QObject *parent) : QAbstractTableModel(parent)
 {
+    connect(this, SIGNAL(dataChanged(QModelIndex, QModelIndex)), this, SLOT(updateTableMetadata(QModelIndex, QModelIndex)));
     _stringMatrix = string_matrix;
-    _rowCount = _stringMatrix.size();
-    _columnCount = maxColumns();
-    _cellCount = cellCount();
+    storeRowSizes();
+    calculateFilledCells();
+    updateTableMetadata(QModelIndex(),QModelIndex());
```

Comment:
It seemed a little awkward to do this via side-effects. What made you decide to do it this way instead?

Reply:
The functions are meant to calculate various information about the CSV data, like row and column count, and number of filled and empty cells, and I've called them in the constructor, so that the information is available to the user ASAP. I'm not sure what exactly is the side effect here. 🤔

As for the changed function to calculate the filled and empty cells, the previous methodology wouldn't have worked because it decided if a cell was filled or not on the basis of existence of its corresponding string matrix element, so it marked a cell as filled even if it had no string, which is not desirable. Now it checks if a cell actually has any text or not to decide if it's filled or not. Now this would be very expensive if it was done for ever cells each time a cell is modified. So I've kept a count of filled cells for each row, and whenever a cell gets modified, the value of filled cells are recomputed only for that row, and the values from the other rows are just added to that value, thereby saving time and resources.

Reply:
A side effect is anything that a function does outside of generating its return value, setting state variables in this case. Having the functions return the value and setting the state from the constructor is generally considered more readable. Not a big deal, just a good practice, especially as you work with more multithreaded code.

Your function names are also a bit messy. The terms `store`, `calculate`, and `update` are all used, but the functions do basically the same thing for 3 different values. You might consider choosing more consistent names next time.

Reply:
Thanks a lot for good advice.

Reply:
Ah, I get it now. Will implement both your recommendations right away. Thanks!

Reply:
Thanks Rohan